File: backend/michicondrias_ecommerce/app/models/ecommerce.py
from sqlalchemy import Column, String, Boolean, ForeignKey, DateTime, Text, Float, Integer
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from app.db.session import Base
import uuid

class Category(Base):
    __tablename__ = "categories"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    name = Column(String, index=True, nullable=False, unique=True)
    description = Column(Text, nullable=True)
    image_url = Column(String, nullable=True)
    is_active = Column(Boolean, default=True)

    subcategories = relationship("Subcategory", back_populates="category", cascade="all, delete-orphan")
    products = relationship("Product", back_populates="category")


class Subcategory(Base):
    __tablename__ = "subcategories"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    name = Column(String, index=True, nullable=False)
    description = Column(Text, nullable=True)
    category_id = Column(String, ForeignKey("categories.id"), nullable=False)
    is_active = Column(Boolean, default=True)

    category = relationship("Category", back_populates="subcategories")
    products = relationship("Product", back_populates="subcategory")


class Product(Base):
    __tablename__ = "products"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    name = Column(String, index=True, nullable=False)
    description = Column(Text, nullable=True)
    price = Column(Float, nullable=False)
    stock = Column(Integer, default=0)
    
    category_id = Column(String, ForeignKey("categories.id"), nullable=True)
    subcategory_id = Column(String, ForeignKey("subcategories.id"), nullable=True)
    
    image_url = Column(String, nullable=True)
    is_active = Column(Boolean, default=True, index=True)
    is_approved = Column(Boolean, default=False, index=True)
    seller_id = Column(String, index=True, nullable=True)
    
    specifications = Column(Text, nullable=True) # JSON or formatted text for tech specs

    category = relationship("Category", back_populates="products")
    subcategory = relationship("Subcategory", back_populates="products")
    reviews = relationship("Review", back_populates="product", cascade="all, delete-orphan")
    order_items = relationship("OrderItem", back_populates="product")

class Review(Base):
    __tablename__ = "product_reviews"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    product_id = Column(String, ForeignKey("products.id"), nullable=False)
    user_id = Column(String, index=True, nullable=False) # From Core Auth
    rating = Column(Integer, nullable=False) # 1-5
    comment = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    product = relationship("Product", back_populates="reviews")

class Order(Base):
    __tablename__ = "orders"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    user_id = Column(String, index=True, nullable=False) # From Core Auth
    total_amount = Column(Float, nullable=False)
    status = Column(String, default="pending") # pending, paid, shipped, delivered, cancelled
    shipping_address = Column(Text, nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    items = relationship("OrderItem", back_populates="order", cascade="all, delete-orphan")

class OrderItem(Base):
    __tablename__ = "order_items"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    order_id = Column(String, ForeignKey("orders.id"), nullable=False)
    product_id = Column(String, ForeignKey("products.id"), nullable=False)
    quantity = Column(Integer, nullable=False)
    price_at_purchase = Column(Float, nullable=False) # Important in case price changes later

    order = relationship("Order", back_populates="items")
    product = relationship("Product")

class Donation(Base):
    __tablename__ = "donations"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()), index=True)
    user_id = Column(String, index=True, nullable=True) # Refers to users.id from core app. Null for anonymous.
    amount = Column(Float, nullable=False)
    currency = Column(String, default="MXN")
    message = Column(Text, nullable=True)
    date = Column(DateTime(timezone=True), server_default=func.now())
    status = Column(String, default="completed") # e.g. "pending", "completed", "failed"

File: mobile/src/types/ecommerce.ts
/**
 * @module types/ecommerce
 * @description Types for the e-commerce domain — product categories, products,
 * reviews, orders, and donations.
 */

// ─── Categories ─────────────────────────────────────────────────────────────────

export interface Category {
    id: string;
    name: string;
    description: string | null;
    image_url: string | null;
}

// ─── Products ───────────────────────────────────────────────────────────────────

export interface Product {
    id: string;
    name: string;
    description: string | null;
    price: number;
    stock: number;
    category_id: string | null;
    category?: Category;
    image_url: string | null;
    is_active: boolean;
    seller_id: string | null;
    specifications: string | null;
    average_rating: number;
    review_count: number;
}

// ─── Reviews ────────────────────────────────────────────────────────────────────

export interface Review {
    id: string;
    product_id: string;
    user_id: string;
    rating: number;
    comment: string | null;
    created_at: string;
}

export interface ReviewCreate {
    rating: number;
    comment?: string;
}

// ─── Orders ─────────────────────────────────────────────────────────────────────

export interface Order {
    id: string;
    user_id: string;
    total_amount: number;
    status: string;
    shipping_address?: string;
    created_at: string;
}

export interface OrderCreate {
    items: { product_id: string; quantity: number }[];
    shipping_address?: string;
}

// ─── Donations ──────────────────────────────────────────────────────────────────

export interface Donation {
    id: string;
    user_id: string | null;
    amount: number;
    currency: string;
    message: string | null;
    date: string;
    status: string;
}

// ─── Constants & Defaults ───────────────────────────────────────────────────────

/** Order status options */
export const ORDER_STATUS_OPTIONS = [
    { label: 'Pendiente', value: 'pending' },
    { label: 'Confirmado', value: 'confirmed' },
    { label: 'Enviado', value: 'shipped' },
    { label: 'Entregado', value: 'delivered' },
    { label: 'Cancelado', value: 'cancelled' },
] as const;

/** Default donation currency */
export const DEFAULT_DONATION_CURRENCY = 'MXN';
